mcp/sensors/csbf_dgps.c

```diff
@@ -406,4 +518,181 @@ void * DGPSMonitor(void * arg)
+            int read_timeout_usec = 1000;
+            read_timeout.tv_sec = 0;
+            read_timeout.tv_usec = read_timeout_usec;
+            setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR | SO_RCVTIMEO, &read_timeout, sizeof(read_timeout));
+            // now we set up the print statement vars
+            // need to cast the socket address to an INET still address
+            struct sockaddr_in *ipv = (struct sockaddr_in *)servinfo->ai_addr;
+            // then pass the pointer to translation and put it in a string
+            inet_ntop(AF_INET, &(ipv->sin_addr), ipAddr, INET_ADDRSTRLEN);
+        }
+        numbytes = recvfrom(
+            sockfd,
+            nmea_buffer,
+            sizeof(nmea_buffer) + 1,
+            0,
+            (struct sockaddr *) &sender_addr,
+            &addr_len);
+        // we get an error everytime it times out, but EAGAIN is ok, other ones are bad.
+        if (numbytes == -1) {
+            err = errno;
+            if (err != EAGAIN) {
+                blast_info("Errno is %d\n", err);
+                blast_info("Error is %s\n", strerror(err));
+                berror(err, "Recvfrom");
+            }
+        } else {
+            // Tokenize str from UDP packet: the GPS receiver we talk to
+            // stuffs all NMEA sentences into a single packet, so we separate
+            // them here
+            // blast_info("Received UDP packet from GNSS module\n");
+            char* pTok = NULL;
+            char* pSave = NULL;
+            pTok = strtok_r(nmea_buffer, "\r\n", &pSave);
+            while (pTok != NULL) {
+                if (!CSBFGPSverifyChecksum(pTok, strlen(pTok))) {
+                    blast_err("checksum failed");
+                } else {
+                    if (NMEA_CHATTER) {
+                        blast_info("Parsing NMEA sentence from UDP...");
+                    }
+                    // Select any handlers that match the sentence
+                    for (nmea_handler_t *handler = handlers; handler->proc; handler++) {
+                        if (!strncmp(pTok, handler->str, (int) strlen(handler->str) - 1)) {
+                            handler->proc(pTok);
+                        }
+                    }
+                }
+                pTok = strtok_r(NULL, "\r\n", &pSave);
+            }
+        }
+    }
+    freeaddrinfo(servinfo);
+    close(sockfd);
+    return NULL;
+}
+
+/**
+ * @brief Multiple methods of receiving NMEA-formatted GNSS data run
+ * concurrently.
+ * @details NMEA messages are parsed from a dedicated serial port and UDP
+ * packets. Access to the data used by the rest of the program is mutexed,
+ * to avoid race conditions when serial/UDP messages arrive simultaneously.
+ * As a result, the last message that comes in "wins".
+ */
+void StartDGPSmonitors(void)
+{
+    // Dedicated serial port
+    pthread_t DGPSserialThread;
+    pthread_create(&DGPSserialThread, NULL, DGPSmonitorSerial, NULL);
+    pthread_detach(DGPSserialThread);
```

Comment:
this detaching just good practice?

Reply:
Yes, but for our case is mostly pedantry, since we typically do not purposefully end threads aside from ending the entire program